docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+```go
+type ISecureStorage interface {
+  List() []string
+
+  Get(name string) (SecureItem, error)
+  Set(item SecureItem) error
+  Remove(name string) error
+}
+```
+
+### Secure Item
+
+A *Secure Item* is a structured data object designed for storing any type of data within a *Secure Storage* instance.
+In the context of this ADR, the `Blob` field of a `SecureItem` represents the serialized form of a `CryptoProvider`.
+
+_Note:_ The encoder/decoder of the `Blob` field will be dependent on the `CryptoProvider` implementation, a `SecureItem`
+has no knowledge of the underlying data structure.
+
+```go
+type ISecureItemMetadata interface {
+  Type()   TypeUUID     // Relates to the corresponding provider
+  Name()   string
+}
+
+type ISecureItem interface {
+  ISecureItemMetadata
+  Bytes() []byte
+}
+```
+
+##### Keyring
+
+The `Keyring` interface serves as a central hub for managing `CryptoProviders` and `SecureStorage` implementations.
+The main difference with the current `Keyring` interface is that it adds the ability to register several `SecureStorage` (this is, several "storing backends") implementations and the
+_values_ are abstracted as `SecureItem` instead of `Record`.
+
+```go
+type IKeyring interface {
+  RegisterCryptoProviderBuilder(typeUUID TypeUUID, builder CryptoProviderBuilder)
+  RegisterAndLoadStorageProvider(typeUUID TypeUUID, provider StorageProvider)
+
+  ListStorageProviders() ([]IStorageProvider, error)
+  ListCryptoProviders() ([]ICryptoProvider, error)
+
+  List() ([]SecureItemMetadata, error)
+  
+  GetCryptoProvider(ItemId) (CryptoProvider, error)
+}
+```
+
+#### **Wallet**
+
+The `Wallet` interface contains the blockchain specifics logics for converting PubKeys into addresses.
+It also serves as a convenient API for:
+
+* Signing and Verifying messages
+* Generating addresses out of keys
+
+_Note:_ Each Wallet implementation should provide the logic to map addresses and PubKeys.
+
+```go
+type Wallet interface {
+	Init(Keyring)
+	GetSigner(address string) Signer
+	GetVerifier(address string) Verifier
+	Generate() string
+}
+```
+
+#### Additional components
+
+##### **Blob**
+
+This is a wrapper for the widely used `[]byte` type that is used when handling binary data.
+Since handling sensitive information is a common task in crypto algorithms, the objective is to provide some extra security capabilities around such type as:
+
+* Zeroing values after a read operation.
+* Proper data handling.
+
+#### **Pub/Priv Key**
+
+PubKey and PrivKey are wrappers to store the corresponding keys bytes.
+The corresponding implementations will have the proper security measures to handle the keys mentioned at the beginning of this document.
+
+```mermaid
+classDiagram
+  PubKey <|-- PrivKey
+  PubKey : Address() string
+  PubKey : Key 
+
+  PrivKey : PubKey() PubKey
+  PrivKey : key
+```
+
+Base key interface (common to private and public keys)
+
+```go
+type BaseKey interface {
```

Comment:
What's the use of BaseKey? Applications should be able to just use PubKey instead.

Reply:
Yes but take into account that in the proposed design, `Pubkey` extends `Basekey`, same with `Privkey` with the advantage that in the latter we can extend it to have data zeroing or 'references' to remote privkeys without having it in memory.